src/backend/joanie/core/models/products.py

```diff
@@ -138,6 +145,43 @@ def __str__(self):
         return f"{self.product}: {self.position} / {self.course}"
 
 
+def on_add_course_runs_to_product_course_relation(action, instance, pk_set, **kwargs):
+    """
+    Signal triggered when course runs are added to a product course relation.
+    Some checks are processed before course runs are linked to product course relation.
+    """
+    if action != "pre_add":
+        return
+
+    if isinstance(instance, ProductCourseRelation):
+        relations = [instance]
+        course_runs_set = pk_set
+    else:
+        relations = ProductCourseRelation.objects.filter(pk__in=pk_set).select_related(
+            "course__course_runs"
+        )
+        course_runs_set = {instance.pk}
+
+    for relation in relations:
+        # Check that course runs relies on the relation course
+        if relation.course.course_runs.filter(pk__in=course_runs_set).count() != len(
```

Comment:
you probably need a distinct to avoir counting a course run twice :thinking: A test could help confirm this intuition :thinking:

Reply:
From my understanding of your comment here is my analysis :

The executed query for

```python
relation.course.course_runs.filter(pk__in=course_runs_set).count()
```
is the following : 

```SQL
SELECT COUNT(*) AS "__count"
FROM "joanie_course_run"
WHERE (
  "joanie_course_run"."course_id" = <COURSE_ID>
  AND "joanie_course_run"."id" IN (<COURSE_RUNS_PK_SET>)
);
```

In this context, I don't think we could count a course run twice.
Did I miss a point from your comment?